src/identity/configuration/KeyGeneratingIdpConfigurationFactory.ts

```diff
@@ -0,0 +1,107 @@
+/* eslint-disable id-length */
+/* eslint-disable max-len */
+/* eslint-disable @typescript-eslint/naming-convention */
+import { randomBytes } from 'crypto';
+import { JWK } from 'node-jose';
+import type { Adapter, Configuration } from 'oidc-provider';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import { getLoggerFor } from '../../logging/LogUtil';
+import type { KeyValueStore } from '../storage/KeyValueStore';
+import type { StorageAdapterFactory } from '../storage/StorageAdapterFactory';
+import { IdpConfigurationFactory } from './IdpConfigurationFactory';
+
+/**
+ * An IdP Configuration Factory that generates and saves keys
+ * to the provided key value store.
+ */
+export class KeyGeneratingIdpConfigurationFactory extends IdpConfigurationFactory {
+  private readonly storageAdapterFactory: StorageAdapterFactory;
+  private readonly baseUrl: string;
+  private readonly store: KeyValueStore;
+  private readonly logger = getLoggerFor(this);
+
+  public constructor(
+    storageAdapterFactory: StorageAdapterFactory,
+    baseUrl: string,
+    store: KeyValueStore,
+  ) {
+    super();
+    this.storageAdapterFactory = storageAdapterFactory;
+    this.baseUrl = baseUrl;
+    this.store = store;
+  }
+
+  private getJwksKey(): ResourceIdentifier {
+    return { path: `${this.baseUrl}/idp/jwks` };
+  }
+
+  // There is a typing difficulty with JSONWebKeySet, thus the "any"
+  private async generateJwks(): Promise<any> {
+    // Check to see if the keys are already saved
+    const jwks = await this.store.get(this.getJwksKey());
+    if (jwks) {
+      return jwks;
+    }
+    // If they are not, generate and save them
+    const keystore = JWK.createKeyStore();
+    await keystore.generate('RSA');
+    const newJwks = keystore.toJSON(true);
+    await this.store.set(this.getJwksKey(), newJwks);
+    return newJwks;
+  }
+
+  private getCookieSecretKey(): ResourceIdentifier {
+    return { path: `${this.baseUrl}/idp/cookie-secret` };
+  }
+
+  private async generateCookieKeys(): Promise<string[]> {
+    // Check to see if the keys are already saved
+    const cookieSecret = await this.store.get(this.getCookieSecretKey());
+    if (Array.isArray(cookieSecret)) {
+      return cookieSecret;
+    }
+    // If they are not, generate and save them
+    const newCookieSecret = [ randomBytes(64).toString('hex') ];
+    await this.store.set(this.getCookieSecretKey(), newCookieSecret);
+    return newCookieSecret;
+  }
+
+  public async createConfiguration(): Promise<Configuration> {
+    const jwks = await this.generateJwks();
+    const cookieKeys = await this.generateCookieKeys();
+
+    // Aliasing the "this" variable is an anti-pattern that is better served by using
+    // arrow functions. Unfortunately, the adapter function MUST be a named function
+    // See https://github.com/panva/node-oidc-provider/issues/799
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const that = this;
```

Comment:
But we just need `storageAdapterFactory`, right?

Reply:
Ah true